### ISSUE TITLE
[Bug] v.in.ogr crashes when importing GPKG layers (MacOS 10.14/GDAL 3.1.0)

### ISSUE THREAD

**Issue by Veence** [OSGeo/grass#610]
Importing GPKG with v.in.ogr crashes – this has been going on for a while now:
```
Air > ogrinfo --formats
Supported Formats:
  […]
  GPKG -raster,vector- (rw+vs): GeoPackage
```

```
ogrinfo Mentions_par_feuille.gpkg
INFO: Open of `Mentions_par_feuille.gpkg' using driver `GPKG' successful.
1: Mentions par feuille (Multi Polygon)
```

```
GRASS GIS 7.8.3 > v.in.ogr -e --overwrite input=Mentions_par_feuille.gpkg output=Mentions_par_feuille
Segmentation fault
```

I’m considering trying and compile Grass with the -g debug flag and tracing what's wrong, if that can help.

**Comment by nilason** [comment]
I could reproduce this with macports build of 7.8.3, with crash report:

```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   ???                           	000000000000000000 0 + 0
1   libsqlite3.dylib              	0x00007fff78f6b881 dbMallocRawFinish + 17
2   libsqlite3.dylib              	0x00007fff78e84666 sqlite3FindFunction + 886
3   libsqlite3.dylib              	0x00007fff78e84b20 sqlite3CreateFunc + 496
4   libsqlite3.dylib              	0x00007fff78e857b7 sqlite3_create_function_v2 + 487
5   libspatialite.7.dylib         	0x00000001031a07e8 register_spatialite_sql_functions + 68
6   libspatialite.7.dylib         	0x00000001031ddf2a spatialite_init_ex + 52
7   libgdal.26.dylib              	0x0000000101371ce2 OGRSQLiteBaseDataSource::InitNewSpatialite() + 194
8   libgdal.26.dylib              	0x0000000101183b62 GDALGeoPackageDataset::InstallSQLFunctions() + 34
9   libgdal.26.dylib              	0x000000010116e13f GDALGeoPackageDataset::OpenOrCreateDB(int) + 79
10  libgdal.26.dylib              	0x0000000101172da0 GDALGeoPackageDataset::Open(GDALOpenInfo*) + 1792
11  libgdal.26.dylib              	0x00000001011906ca OGRGeoPackageDriverOpen(GDALOpenInfo*) + 106
12  libgdal.26.dylib              	0x0000000100d02d26 GDALOpenEx + 2790
13  v.in.ogr                      	0x00000001001c50f3 main + 2759
14  libdyld.dylib                 	0x00007fff79d073d5 start + 1
```

but not with my dev set-up, where I don´t patch up gdal. I need to look into this more thoroughly, but maybe it's related with that patch?
I've intended to ask you why the patch is needed (see also https://github.com/OSGeo/grass/issues/326#issuecomment-619612799). If you test without the patch, can you reproduce the segfault?

**Comment by Veence** [comment]
Ah, thanks for the clue! I’m going to check this later this afternoon.

**Comment by Veence** [comment]
I just tested with the patch disabled. I get the same crash. I update libspatialite to the latest commit, btw.

**Comment by nilason** [comment]
Could you please attach a (any) gpkg file which is causing the issue? Maybe we are not seeing the same problems.

**Comment by Veence** [comment]
My trace log is exactly the same as yours though

**Comment by nilason** [comment]
After some further testing I can only repeat that this isn't an issue with my standard build (configure; make etc.) from outside macports, even if I use almost everything from macports:


**_MacPorts:_**

```
$ grass78 --config
x86_64-apple-darwin18.7.0
./configure  --prefix=/opt/local --prefix=/opt/local/share --exec-prefix=/opt/local --with-opencl=yes --with-bzlib --with-bzlib-includes=/opt/local/include --with-bzlib-libs=/opt/local/lib --with-fftw-includes=/opt/local/include --with-fftw-libs=/opt/local/lib --with-freetype --with-freetype-includes=/opt/local/include/freetype2 --with-freetype-libs=/opt/local/lib --with-gdal=/opt/local/bin/gdal-config --with-pdal=/opt/local/bin/pdal-config --with-tiff-includes=/opt/local/include --with-tiff-libs=/opt/local/lib --with-png-includes=/opt/local/include --with-png-libs=/opt/local/lib --with-geos=/opt/local/bin/geos-config --with-liblas-includes=/opt/local/include/liblas --with-liblas-libs=/opt/local/lib --without-postgres --with-pthread --with-cairo --with-cairo-includes=/opt/local/include/cairo --with-cairo-libs=/opt/local/lib --with-blas=no --with-lapack=no --with-readline --with-readline-includes=/opt/local/include/readline --with-readline-libs=/opt/local/lib --with-proj-includes=/opt/local/lib/proj6/include --with-proj-libs=/opt/local/lib/proj6/lib --with-proj-share=/opt/local/lib/proj6/share/proj --with-wxwidgets=/opt/local/Library/Frameworks/wxWidgets.framework/Versions/wxWidgets/3.1/bin/wx-config --with-opengl=osx --with-postgres-includes=/opt/local/include/postgresql12 --with-postgres-libs=/opt/local/lib/postgresql12 --with-postgres=yes --with-sqlite --with-sqlite-includes=/opt/local/include --with-sqlite-libs=/opt/local/lib --with-zstd --with-zstd-includes=/opt/local/include --with-zstd-libs=/opt/local/lib
/usr/bin/clang
/opt/local/share/grass78
exported
7.8.3

GRASS 7.8.3 [...] > g.version -rge
version=7.8.3
date=2020
revision=exported
build_date=2020-05-11
build_platform=x86_64-apple-darwin18.7.0
build_off_t_size=8
libgis_revision=2020-05-11T16:54:28+00:00
libgis_date=2020-05-11T16:54:28+00:00
proj=6.3.2
gdal=3.1.0
geos=3.8.1
sqlite=3.31.1
```

**_My dev set-up:_**

```
$ ./bin.x86_64-apple-darwin18.7.0/grass78 --config
x86_64-apple-darwin18.7.0
./configure  --with-iconv-libs=/usr/lib --with-freetype --with-freetype-includes=/opt/local/include/freetype2 --with-freetype-libs=/opt/local/lib --with-gdal=/opt/local/bin/gdal-config --with-geos=/opt/local/bin/geos-config --with-proj --with-proj-includes=/opt/local/lib/PROJ6/include --with-proj-libs=/opt/local/lib/PROJ6/lib --with-proj-share=/opt/local/lib/PROJ6/share/proj --with-png-includes=/opt/local/include --with-png-libs=/opt/local/lib --with-tiff-includes=/opt/local/include --with-tiff-libs=/opt/local/lib --with-postgres=yes --with-postgres-includes=/opt/local/include/postgresql10 --with-postgres-libs=/opt/local/lib/postgresql10 --with-liblas-includes=/opt/local/include/liblas--with-liblas-libs=/opt/local/lib --without-mysql --with-pthread --with-cairo --with-cairo-includes=/opt/local/include/cairo --with-cairo-libs=/opt/local/lib --with-cairo-ldflags=-lcairo --with-blas=no --with-lapack=no --with-bzlib --with-bzlib-includes=/opt/local/include --with-bzlib-libs=/opt/local/lib --with-zstd --with-zstd-includes=/opt/local/include --with-zstd-libs=/opt/local/lib --with-sqlite --with-sqlite-libs=/opt/local/lib --with-sqlite-includes=/opt/local/include --with-fftw-includes=/opt/local/include --with-fftw-libs=/opt/local/lib --with-readline --with-readline-includes=/opt/local/include/readline --with-readline-libs=/opt/local/lib --with-opengl=aqua
gcc
/Volumes/dev/grass/dist.x86_64-apple-darwin18.7.0
550abd175
7.8.4dev

GRASS 7.8.4dev [...] > g.version -rge	
version=7.8.4dev
date=2020
revision=550abd175
build_date=2020-05-11
build_platform=x86_64-apple-darwin18.7.0
build_off_t_size=8
libgis_revision=8e4dab571
libgis_date=2020-05-11T16:26:18+00:00
proj=6.3.2
gdal=3.1.0
geos=3.8.1
sqlite=3.31.1
```

You're right, removing the gdal patch doesn't make any difference.

**Comment by Veence** [comment]
What about spatialite?

**Comment by nilason** [comment]
```
$ port -v installed spatialite
The following ports are currently installed:
  spatialite @4.99.20200511_0+proj6 (active) platform='darwin 18' archs='x86_64' date='2020-05-11T14:07:43+0200'

$ port deps grass7
Full Name: grass7 @7.8.3_1+postgresql12+python38
Build Dependencies:   pkgconfig
Library Dependencies: bzip2, freetype, fftw-3, gdal, pdal, geos, tiff, libpng, liblas, cairo, readline, proj, python38, py38-Pillow, py38-numpy, py38-six, postgresql12
Runtime Dependencies: py38-psycopg2

$ port dependents spatialite
gdal depends on spatialite

```

**Comment by Veence** [comment]
Ok. I’m trying to get a complete readout of the backtrace. So far, I have in Xcode everything but the innermost libsqlite3 call. The rest is here complete with source and variables, but I still don’t know why lldb won’t pick up the .dSYM directory associated with libsqlite3.0.dylib. I’ll pick up tomorrow. Thanks for your involvement!

**Comment by Veence** [comment]
As an addendum, I just found out that there are TWO libsqlite pulled by v.in.org. The Macports one AND the system one. I wonder if that doesn’t cause the crash, because who know which one is called?

**Comment by Veence** [comment]
OK. My version of libgdal pulls in /System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa which itself pulls the system libsqlite3 in (because it depends on CoreData).
Has your own libgdal also this dependency? If not, what variants did you choose to build it?

**Comment by nilason** [comment]
```
$ port installed gdal
The following ports are currently installed:
  gdal @3.1.0_0 (active)
```

```
$ otool -L /opt/local/share/grass78/bin/v.in.ogr 
/opt/local/share/grass78/bin/v.in.ogr:
	/System/Library/Frameworks/OpenCL.framework/Versions/A/OpenCL (compatibility version 1.0.0, current version 1.0.0)
	/opt/local/share/grass78/lib/libgrass_gproj.7.8.dylib (compatibility version 7.8.0, current version 7.8.0)
	/opt/local/share/grass78/lib/libgrass_vector.7.8.dylib (compatibility version 7.8.0, current version 7.8.0)
	/opt/local/share/grass78/lib/libgrass_dbmiclient.7.8.dylib (compatibility version 7.8.0, current version 7.8.0)
	/opt/local/share/grass78/lib/libgrass_dbmibase.7.8.dylib (compatibility version 7.8.0, current version 7.8.0)
	/opt/local/share/grass78/lib/libgrass_gis.7.8.dylib (compatibility version 7.8.0, current version 7.8.0)
	/opt/local/lib/libgdal.27.dylib (compatibility version 28.0.0, current version 28.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

```
$ otool -L /opt/local/lib/libgdal.27.dylib
/opt/local/lib/libgdal.27.dylib:
	/opt/local/lib/libgdal.27.dylib (compatibility version 28.0.0, current version 28.0.0)
	/opt/local/lib/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	/opt/local/lib/libqhull.7.dylib (compatibility version 7.0.0, current version 7.3.2)
	/opt/local/lib/libgeos_c.1.dylib (compatibility version 15.0.0, current version 15.3.0)
	/usr/lib/libexpat.1.dylib (compatibility version 7.0.0, current version 8.0.0)
	/opt/local/lib/libgif.4.dylib (compatibility version 6.0.0, current version 6.7.0)
	/opt/local/lib/libCharLS.2.dylib (compatibility version 2.0.0, current version 2.0.0)
	/opt/local/lib/libpng16.16.dylib (compatibility version 54.0.0, current version 54.0.0)
	/opt/local/lib/libzstd.1.dylib (compatibility version 1.0.0, current version 1.4.4)
	/opt/local/lib/proj6/lib/libproj.15.dylib (compatibility version 19.0.0, current version 19.1.0)
	/opt/local/lib/libsqlite3.0.dylib (compatibility version 9.0.0, current version 9.6.0)
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
	/opt/local/lib/libspatialite.7.dylib (compatibility version 9.0.0, current version 9.1.0)
	/opt/local/lib/libpcre.1.dylib (compatibility version 4.0.0, current version 4.12.0)
	/opt/local/lib/libcurl.4.dylib (compatibility version 11.0.0, current version 11.0.0)
	/opt/local/lib/libiconv.2.dylib (compatibility version 9.0.0, current version 9.1.0)
	/opt/local/lib/libxml2.2.dylib (compatibility version 12.0.0, current version 12.10.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
```

**Comment by Veence** [comment]
Thanks. As I suspected, it’s the ECW driver that pulls the system libsqlite3 into the mix. I've eliminated it, but it is still there, so I assume there’s another "leaky" path somewhere. Problem is I have to examine dependents of dependents, etc. and it’s at best very cumbersome. I wonder if there is a tool that can do that automatically.

**Comment by neteler** [comment]
When you find a way, please add it here: https://grasswiki.osgeo.org/wiki/GRASS_Debugging

**Comment by nilason** [comment]
maybe https://sites.google.com/a/puredarwin.org/puredarwin/developers/otool/mach-o-dependencies-overview, for code: https://github.com/PureDarwin/PureDarwin/blob/master/scripts/pd_machviz

**Comment by Veence** [comment]
Tried that. It _seems_ to work, but I can’t be definitive because it's so inefficient that I had to abort it after one hour, with transient error messages popping up every now and then (nothing I could catch given that they went away as fast as they had appeared).

I’m going to write my own tool, probably using Python (or Perl, maybe) instead of that mix of shell/graphviz. But thanks a bunch for the suggestion.

**Comment by nilason** [comment]
> Tried that. It seems to work, but I can’t be definitive because it's so inefficient that I had to abort it after one hour, with transient error messages popping up every now and then (nothing I could catch given that they went away as fast as they had appeared).

...didn't work out of the box for me either, but looked promising...

meanwhile I can confirm too through lldb debugging, that my pure macports build use system `/usr/lib/libsqlite3.dylib` while my "standard" (and working) build use `/opt/local/lib/libsqlite3.0.dylib`. Both libraries are loaded (`(lldb) image list)`, at both times.

**Comment by Veence** [comment]
This seems to jibe with what I surmise, i.e. that the loading of both libraries is the cause of the crash. I will have a busy afternoon, but I’ll try to pick up on that later tonight.

**Comment by nilason** [comment]
> When you find a way, please add it here: https://grasswiki.osgeo.org/wiki/GRASS_Debugging

It is on my todo checklist, for a while, to update debugging instructions using lldb, in particular for mac and also wxpython.

**Comment by Veence** [comment]
Strange, as the tool I just wrote doesn’t detect any libsqlite3.dylib dependency

```
/opt/local/bin/python3.8 /Users/vincent/PycharmProjects/dpdanalyse/dpdanalyse.py -n sqlite /opt/local/share/grass78/bin/v.in.ogr
The following libraries depend on /opt/local/lib/libsqlite3.0.dylib (sqlite):
 + /opt/local/lib/libgdal.27.dylib
 + /opt/local/lib/libspatialite.7.dylib
 + /opt/local/lib/proj6/lib/libproj.15.dylib
```

The only deps found are on the Macport's libsqlite…

**Comment by Veence** [comment]
[https://pasteboard.co/J84ZVSU.png](url)
Lol

**Comment by nilason** [comment]
Nice graph!

> The only deps found are on the Macport's libsqlite…

...how frustrating...

Did you try approaching the issue from "outside" macports? Something in line with what I made to be able to make a working build. Maybe something falls out in the process...

**Comment by nilason** [comment]
but wait, I did a test too with a slightly modified (only to get it work) script following https://stackoverflow.com/a/1517652. The system `/usr/lib/libsqlite3.dylib` and the macports `/opt/local/lib/libsqlite3.0.dylib` library did turn up for both my builds. The problem with that script is that it doesn't tell the calling paths.

**Comment by Veence** [comment]
Thanks for the graph. I have added an extra option to the tool to get rid of the /usr/lib/system libraries, it gives a neater output. But still I don’t see any reference to the system libsqlite3 in the graph, so I think I’ll have to get acquainted with dtrace.

**Comment by Veence** [comment]
Ok, I have it.

First, rather than use Dtrace (which is a very powerful tool of its own, and I will dive into it because it is really worth it), there are environment variables you can set that will cause dyld(1) to print out the name of the libraries it processes. You can find them [here](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/LoggingDynamicLoaderEvents.html).

In short, I did this and… yes, I found grass loaded libsqlite3.dylib. The culprit was… the OpenCL framework. My dependency tracing tool is not 100% finished: it ignores the Frameworks lib right now, so that dependency was not recorded. I will update the code to properly take frameworks into account.

I recompiled grass78 specifying --without-opencl during the configure phase, and it works fine now.

I can safely close this.

**Comment by nilason** [comment]
Great job, thanks!

I knew that in the end there will be a trivial answer to this...
